Add entity names to get_df index_by_series

stephenbernardi opened the pull request:
To finish up CLEWS-15676, this PR converts entity ids to names when index_by_series is chosen.

In addition, another option (index_by_end_date) is added to streamline get_df's output so each series is in a single column and end date is set as the index.

bernardigro committed:
fix source id column addition

bernardigro committed:
update get_df to index by entity name and add index by date functionality

bernardigro committed:
return empty df if multiple GDH series are found

stephenbernardi commented:
tested with the following code:
`import os

from api.client.gro_client import GroClient


API_HOST     = 'api.gro-intelligence.com'
ACCESS_TOKEN = os.environ['GROAPI_TOKEN']
client       = GroClient(API_HOST, ACCESS_TOKEN)

brazil_sugar_prod = client.get_data_series(**{'metric_id': 860032, 
                                              'item_id': 13615, 
                                              'region_id': 1029, 
                                              'partner_region_id': 0, 
                                              'frequency_id': 9})
 

# guangdong cash prices soy meal
x = client.GDH('15851987-287-13322-0-1-128')

for s in brazil_sugar_prod:
    client.add_single_data_series(s)

df = client.get_df(index_by_series=True)
df2 = client.get_df(index_by_end_date=True)
`

here is an example of index_by_series output: 
![image](https://user-images.githubusercontent.com/38086048/86519255-eb00c980-be06-11ea-873c-af180589fe25.png)

here is an example of index_by_end_date output:
![image](https://user-images.githubusercontent.com/38086048/86519265-feac3000-be06-11ea-8498-603d0b71463f.png)

itsjohnward committed:
Merge branch 'development' into CLEWS-15676-add-names-to-index

itsjohnward commented on api/client/gro_client.py:
```diff
@@ -1082,9 +1111,16 @@ def GDH(self, gdh_selection, **optional_selections):
+            self._logger.warn("Multiple series returned for GDH selection")
+            return pandas.DataFrame()
+        for series in gdh_series: self.add_single_data_series(series)
+        entity_names = [series[entity_id.replace('id', 'name')] \
```
There's a couple failing unit tests on this PR.

One of them's related to this line, breaking GDH:

```py
In [4]: client.GDH('860032-274-1215-0-2-9')                                                         
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-4-6dcbd49e07c8> in <module>
----> 1 client.GDH('860032-274-1215-0-2-9')

~/api-client/api/client/gro_client.py in GDH(self, gdh_selection, **optional_selections)
   1118         for series in gdh_series: self.add_single_data_series(series)
   1119         entity_names = [series[entity_id.replace('id', 'name')] \
-> 1120                         for entity_id in DATA_SERIES_UNIQUE_TYPES_ID]
   1121 
   1122         try:

~/api-client/api/client/gro_client.py in <listcomp>(.0)
   1118         for series in gdh_series: self.add_single_data_series(series)
   1119         entity_names = [series[entity_id.replace('id', 'name')] \
-> 1120                         for entity_id in DATA_SERIES_UNIQUE_TYPES_ID]
   1121 
   1122         try:

NameError: free variable 'series' referenced before assignment in enclosing scope
```

nemozen commented on api/client/gro_client.py:
```diff
@@ -874,12 +876,39 @@ def get_df(self, show_revisions=False, index_by_series=False):
+                
+                # add entity names to dataframe
+                series_df = pandas.DataFrame([dict(s) for s in self.get_data_series_list()])
+                df_with_names = self._data_frame.merge(series_df[columns + column_names],
```
add copy=False. merge makes a copy by default, so allowing two copies to exist in memory t the same time would cause  problems.

stephenbernardi replied:
done

nemozen commented on api/client/gro_client.py:
```diff
@@ -874,12 +876,39 @@ def get_df(self, show_revisions=False, index_by_series=False):
+                                                       on=columns)
+                
+                # add unit name
+                unit_lookup = {unit_id:self.lookup('units', unit_id)['name'] \
```
space after `:`

nemozen commented on api/client/gro_client.py:
```diff
@@ -874,12 +876,39 @@ def get_df(self, show_revisions=False, index_by_series=False):
+                
+                #set entity names as index
+                indexed_df = df_with_names.set_index(column_names)
+                indexed_df.index.set_names(column_names, inplace=True)
```
entity names are not unique in general so indexing by names will mix data series. the underlying index should be as before by entity id. and we can just add the names on top MultiIndex.

stephenbernardi replied:
reverted the index by name change

nemozen commented on api/client/gro_client.py:
```diff
@@ -1082,9 +1111,16 @@ def GDH(self, gdh_selection, **optional_selections):
         selection = zip_selections(entity_ids, **optional_selections)
 
-        self.add_single_data_series(selection)
+        gdh_series = self.get_data_series(**selection)
```
we don't want to call get_data series here, as this is at the data retrieval "layer"  while get_data_series is exploration layer.  also don't want to change the semantics of this function... it will return a single series because gdh string is specified that way and this is a selection.

all we would need here is to convert `selection` into `entity names` i.e. get a list of (entity_type, entity_id) from selections and  `entity_names = [self.lookup(entity_type, entity_id) ....]`

nemozen replied:
that said, see comment in get_df(),  entity names are not unique in general so we can't really ise them as the index to get series, we need ids and then have names as just an option output.  so this function can remain unchanged.

stephenbernardi replied:
reverted the GDH changes

nemozen commented on api/client/gro_client.py:
```diff
@@ -866,6 +866,8 @@ def get_df(self, show_revisions=False, index_by_series=False):
             See https://developers.gro-intelligence.com/data-point-definition.html
             If index_by_series is set, the dataframe is indexed by series.
             See https://developers.gro-intelligence.com/data-series-definition.html
+            If index_by_end_date is set, the dataframe is indexed by end date, with the 
```
let's not combine two features in one PR. if we index by date do it in a separate pr. also in this implementation, if index_by_series is false then index_by_end_date has no effect which is confusing...

nemozen reviewed (changes requested):
1) we need to keep the underlying index to be by entity id and use a multindex for names and 2) the date change should be in a separate pr and either work with/without index_by_series, or not be a separate option but implicit in the latter

bernardigro committed:
revert index by end date changes

bernardigro committed:
revert GDH changes

bernardigro committed:
add entity names to get_df index by series output

bernardigro committed:
move entity names to column headers

stephenbernardi commented:
I will do the end_date indexing feature in a different PR as suggested, so that part has been removed. I reverted the changes to the index so it should still be indexed by ids. And the entity names are now in the dataframe column headers. @nemozen @itsjohnward please take another look. thanks

itsjohnward committed:
Merge branch 'development' into CLEWS-15676-add-names-to-index

itsjohnward committed:
fix: resolve failing tests by commenting out index cols lines

itsjohnward commented on api/client/gro_client.py:
```diff
@@ -877,8 +877,32 @@ def get_df(self, show_revisions=False, index_by_series=False):
+                indexed_df.drop(name_columns, axis=1, inplace=True)
+
+                # indexed_df = pandas.concat([indexed_df], keys=indexed_cols, axis=1)
+                # indexed_df.columns.names = name_columns + ['entities']
```
I commented out these two lines because they were causing tests to fail. Without them, it seems to work, but I'm still trying to understand what the intent here is

stephenbernardi replied:
that's where the entity names are added to the column headers. removing these lines defeats the purpose of this PR, or we need to find a different way to accomplish it

itsjohnward committed:
Merge remote-tracking branch 'origin/development' into CLEWS-15676-add-names-to-index

itsjohnward committed:
feat: add entity names to index

itsjohnward committed:
test: fix outdated index-zipping unit test

itsjohnward committed:
fix: remove entity names from index

itsjohnward committed:
feat: add index_by_names option to gdh

itsjohnward committed:
fix: do not make include_names the default

itsjohnward committed:
Merge remote-tracking branch 'origin/development' into CLEWS-15676-add-names-to-index

itsjohnward committed:
feat: index gdh by names always

itsjohnward committed:
fix: intersect index names with DATA_SERIES_UNIQUE_TYPES_ID

itsjohnward commented:
@stephenbernardi I made some changes, so now:

1. `get_df()` has an option to add names as columns, but the index remains unchanged
2. `GDH()` uses names as its index instead of ids

Does that accomplish what you were after here, or did I miss something?

itsjohnward committed:
fix: add empty df check to include_names

stephenbernardi commented:
> @stephenbernardi I made some changes, so now:
> 
> 1. `get_df()` has an option to add names as columns, but the index remains unchanged
> 2. `GDH()` uses names as its index instead of ids
> 
> Does that accomplish what you were after here, or did I miss something?

@itsjohnward thanks! This looks pretty close. I think get_df has the desired output. But the index_by_series option can be independent of the include_names option. I don't think we always need to return the names when indexing by series.

stephenbernardi commented:
also, we should include unit name in the output I think

bernardigro committed:
remove names from index by series option

bernardigro committed:
add unit to include_names

jli commented:
Hey @stephenbernardi, it looks like JohnW had approved these changes, but this wasn't merged.

Is this still something we'd like to do? I see the associated Jira is still open.

stephenbernardi commented:
> Hey @stephenbernardi, it looks like JohnW had approved these changes, but this wasn't merged.
> 
> Is this still something we'd like to do? I see the associated Jira is still open.

@jli yes, I still think these changes would be useful. I think @nemozen needs to review before merging

nemozen commented:
> > Hey @stephenbernardi, it looks like JohnW had approved these changes, but this wasn't merged.
> > Is this still something we'd like to do? I see the associated Jira is still open.
> 
> @jli yes, I still think these changes would be useful. I think @nemozen needs to review before merging

yes had requested some changes back in july .. will take another pass shortly

jli commented:
Just a thought: should get_data_series also have this option? If so, should we try to implement this for both methods in a generic way?

nemozen commented on groclient/client.py:
```diff
@@ -874,13 +877,24 @@ def get_df(self, show_revisions=False, index_by_series=False):
-                return indexed_df.sort_index()
-        return self._data_frame
+
+        df = self._data_frame.copy()
```
the data frame can get pretty large so if we make a copy, the client may run out of memory.. How about we just do the name based index in place.

nemozen replied:
fixed

nemozen reviewed (commented):
looks good, just one comment re: memory usage

nemozen commented:
> Just a thought: should get_data_series also have this option? If so, should we try to implement this for both methods in a generic way?

get_data_series already includes names in the response https://developers.gro-intelligence.com/api.html#groclient.GroClient.get_data_series

jli commented on groclient/client.py:
```diff
@@ -850,7 +850,10 @@ def get_top(self, entity_type, num_results=5, **selection):
-    def get_df(self, show_revisions=False, index_by_series=False):
+    def get_df(self,
+               show_revisions=False,
+               include_names=False,
```
It would be a bit safer to add `include_names` as the last argument. Adding it here could break existing code.

For example, the meaning of `client.get_df(True, True)` would change from setting index_by_series=True to setting index_by_series=False and include_names=True.

nemozen replied:
done

jli commented on groclient/client.py:
```diff
@@ -850,7 +850,10 @@ def get_top(self, entity_type, num_results=5, **selection):
-    def get_df(self, show_revisions=False, index_by_series=False):
+    def get_df(self,
+               show_revisions=False,
+               include_names=False,
```
Also, could you update the docstring with a brief description of what include_names does (around line 871)?

nemozen replied:
done

nemozen committed:
don't make a copy in get_df

nemozen committed:
Merge branch 'development' into CLEWS-15676-add-names-to-index

nemozen committed:
fix merge conflict resolution error

nemozen committed:
put new param of get_df() last for backward compatibility, add parameters' descr to doscstring

tested mem usage of names, pretty consistently around 5.2x, e.g.

>>> from groclient import client
>>> import os
>>> c = client.GroClient('api.gro-intelligence.com', os.environ['GROAPI_TOKEN'])
>>> c.add_data_series(item = "Linseed", region = "Ethiopia")
>>> c.add_data_series(item = "Linseed", region = "United States")
>>> df = c.get_df()
>>> import sys
>>> sys.getsizeof(df)
45944
>>> df = c.get_df(include_names=True)
>>> sys.getsizeof(df)
239240

nemozen commented:
Updated to not grow memory usage by making a copy with each get_df().

Note that include_names makes the df use 5x more memory (see
https://github.com/gro-intelligence/api-client/pull/236/commits/86c65fcc8ab144f0f94e428213d4689eb2401599 ) so it should be used with caution.
